### PR TITLE
Implement conditional requests using ETag

### LINKED ISSUE #104: Provide possibility to use conditional requests (using ETag)
It would be nice to be able to supply an `If-None-Match` or `If-Modified-Since` header and receive `ETag` and `Last-Modified` headers in order to avoid hitting the rate limit too fast.

See https://developer.github.com/v3/#conditional-requests

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #104

Hey @softprops, here's my initial implementation. It's just an initial proposal and I'm happy to make adjustments as you see fit.  What I'm after is I'd like to have this functionality (in one shape or other) in hubcaps without having to maintain a fork of hubcaps, so any solution that avoids that is good for me.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -377,10 +382,18 @@ where
...
         };
         let instance = self.clone();
+        let uri2 = uri.clone();
         let body2 = body.clone();
         let method2 = method.clone();
         let response = url.map_err(Error::from).and_then(move |url| {
             let mut req = Request::builder();
+
+            if method2 == Method::GET {
+                if let Ok(etag) = lookup_etag(&uri2) {
```
Comment (softprops) [inline src/lib.rs]: I like what you're going for here. I'm wondering if you had thoughts on making this an opt in feature and if what those opt in interfaces might look like. I might consider this suprising behavior in environments that restrict where you can write to disk and if you can write to disk.

**Comment by softprops** [comment]
Like the builder pattern.

**Comment by dwijnand** [comment]
Could you give an example of what you mean?

**Comment by dwijnand** [comment]
WDYT?

**Comment by dwijnand** [comment]
This impacts paging, I don't understand why. When the call is made to content no in the cache, all the content is fetched from GitHub, unfolding the pages. On the second call only the first page is uncached.

Do you know why that might be, @softprops?

**Comment by softprops** [comment]
apologies for dragging on this on @dwijnand I'm in the process of moving into a new apartment. I'll try to take a closer look soon. I haven't forgot about this pull!

**Comment by softprops** [comment]
I'm going to opt for iteration here. I'm trying to get better at not lerting pulls go stale. I'd be open to follow ups if needed. I want to roll out some of the recent changes so I'm trying to get as much as I can in. We can iterate on this

**Comment by dwijnand** [comment]
> This impacts paging, I don't understand why. When the call is made to content not in the cache, all the content is fetched from GitHub, unfolding the pages. On the second call only the first page is uncached.

Fixed in #183.

**Pull request merged**